Images/clearlinux

avalluri opened the pull request:
This change allows us to build clearlinux based docker images. We also use clearlinux for building the source.

okartau commented:
- lvmetad is misspelled as lvmeated in comments part, twice
- should we add comment stating why we add -Zn to lvcreate ? The zeroing is called few lines below, but it is not obvious what Zn means, without looking up lvcreate options.

avalluri commented:
> * lvmetad is misspelled as lvmeated in comments part, twice
> * should we add comment stating why we add -Zn to lvcreate ? The zeroing is called few lines below, but it is not obvious what Zn means, without looking up lvcreate options.

Thanks, @okartau. I fixed them now. Can you check if the comment needs any changes?

pohly commented:
I experimented a bit more with selectively disabling the cache via this extra variable. This works rather well. When building the three images, setting up the build environment is done the same way three times, but because it uses the exact same commands, it actually only gets done once and then the following two images reuse the cached result.

The actual runtime images don't share the base because they all use different `swupd update && swupd bundle add` calls. If we ever find that the `swupd update` part in `pmem-ns-init` becomes significant, then we can split up the commands in the other two images to enable sharing.

@avalluri please fold these changes into your PR:
```
diff --git a/Makefile b/Makefile
index 9476a055..23a82bc1 100644
--- a/Makefile
+++ b/Makefile
@@ -59,8 +59,14 @@ push-images: push-pmem-csi-driver-image push-pmem-ns-init-image push-pmem-vgm-im
 pmem-csi-driver pmem-vgm pmem-ns-init:
 	GOOS=linux go build -ldflags '-X main.version=${VERSION}' -a -o _output/$@ ./cmd/$@
 
+# The default is to refresh the base image once a day.
+# This is achieved by passing a fake variable that changes its value once a day.
+# A CI system that produces production images should instead use
+# make  BUILD_IMAGE_ID=<some unique number>
+BUILD_IMAGE_CACHE=$(shell date +%Y-%m-%d)
+
 build-%-image:
-	docker build ${BUILD_ARGS} ${IMAGE_BUILD_ARGS} -t $(IMAGE_TAG) -f ./cmd/$*/Dockerfile . --label revision=${VERSION}
+	docker build --pull --build-arg CACHEBUST=$(BUILD_IMAGE_ID) $(BUILD_ARGS) $(IMAGE_BUILD_ARGS) -t $(IMAGE_TAG) -f ./cmd/$*/Dockerfile . --label revision=$(VERSION)
 
 push-%-image: build-%-image
 	docker push $(IMAGE_TAG)
diff --git a/cmd/pmem-csi-driver/Dockerfile b/cmd/pmem-csi-driver/Dockerfile
index abedeb8f..d6abe325 100644
--- a/cmd/pmem-csi-driver/Dockerfile
+++ b/cmd/pmem-csi-driver/Dockerfile
@@ -4,9 +4,10 @@ ARG VERSION
 ARG NDCTL_VERSION
 ARG NDCTL_CONFIGFLAGS
 ARG NDCTL_BUILD_DEPS
+ARG CACHEBUST
+
 #pull dependencies required for downloading and building libndctl
-RUN swupd update &&  \
-    swupd bundle-add ${NDCTL_BUILD_DEPS} go-basic-dev
+RUN swupd update && swupd bundle-add ${NDCTL_BUILD_DEPS} go-basic-dev && rm -rf /var/lib/swupd
 
 WORKDIR /
 RUN curl -LO https://github.com/pmem/ndctl/archive/v${NDCTL_VERSION}.tar.gz
@@ -27,11 +28,10 @@ RUN mv ./_output/pmem-csi-driver /go/bin/
 
 # build clean container
 FROM clearlinux:base
+ARG CACHEBUST
 LABEL maintainers="Intel"
-LABEL description="Pmem CSI Driver"
-RUN swupd update; \
-    swupd bundle-add binutils file xfsprogs storage-utils; \
-    rm -rf /var/lib/swupd/*
+LABEL description="PMEM CSI Driver"
+RUN swupd update && swupd bundle-add binutils file xfsprogs storage-utils && rm -rf /var/lib/swupd
 
 # move required binaries and libraries to clean container
 COPY --from=build /usr/lib/libndctl* /usr/lib/
diff --git a/cmd/pmem-ns-init/Dockerfile b/cmd/pmem-ns-init/Dockerfile
index 2f436521..66c042d4 100644
--- a/cmd/pmem-ns-init/Dockerfile
+++ b/cmd/pmem-ns-init/Dockerfile
@@ -4,9 +4,10 @@ ARG VERSION
 ARG NDCTL_VERSION
 ARG NDCTL_CONFIGFLAGS
 ARG NDCTL_BUILD_DEPS
+
 #pull dependencies required for downloading and building libndctl
-RUN swupd update &&  \
-    swupd bundle-add ${NDCTL_BUILD_DEPS} go-basic-dev
+ARG CACHEBUST
+RUN swupd update && swupd bundle-add ${NDCTL_BUILD_DEPS} go-basic-dev && rm -rf /var/lib/swupd
 
 WORKDIR /
 RUN curl -LO https://github.com/pmem/ndctl/archive/v${NDCTL_VERSION}.tar.gz
@@ -27,11 +28,12 @@ RUN mv ./_output/pmem-ns-init /go/bin/
 
 # build clean container
 FROM clearlinux:base
-RUN swupd update; \
-    rm -rf /var/lib/swupd/*
+
+ARG CACHEBUST
+RUN swupd update && rm -rf /var/lib/swupd
 
 LABEL maintainers="Intel"
-LABEL description="Pmem CSI Driver namesapce initializer"
+LABEL description="PMEM CSI Driver namesapce initializer"
 # move required binaries and libraries to clean container
 COPY --from=build /usr/lib/libndctl* /usr/lib/
 COPY --from=build /usr/lib/libdaxctl* /usr/lib/
diff --git a/cmd/pmem-vgm/Dockerfile b/cmd/pmem-vgm/Dockerfile
index cf78266d..54e56163 100644
--- a/cmd/pmem-vgm/Dockerfile
+++ b/cmd/pmem-vgm/Dockerfile
@@ -4,9 +4,10 @@ ARG VERSION
 ARG NDCTL_VERSION
 ARG NDCTL_CONFIGFLAGS
 ARG NDCTL_BUILD_DEPS
+
 #pull dependencies required for downloading and building libndctl
-RUN swupd update &&  \
-    swupd bundle-add ${NDCTL_BUILD_DEPS} go-basic-dev
+ARG CACHEBUST
+RUN swupd update && swupd bundle-add ${NDCTL_BUILD_DEPS} go-basic-dev && rm -rf /var/lib/swupd
 
 WORKDIR /
 RUN curl -LO https://github.com/pmem/ndctl/archive/v${NDCTL_VERSION}.tar.gz
@@ -27,12 +28,12 @@ RUN mv ./_output/pmem-vgm /go/bin/
 
 # build clean container
 FROM clearlinux:base
-RUN swupd update; \
-    swupd bundle-add storage-utils; \
-    rm -rf /var/lib/swupd/*
+
+ARG CACHEBUST
+RUN swupd update && swupd bundle-add storage-utils && rm -rf /var/lib/swupd
 
 LABEL maintainers="Intel"
-LABEL description="Pmem CSI driver volume group manager"
+LABEL description="PMEM CSI driver volume group manager"
 #RUN apk add --update kmod eudev util-linux libuuid lvm2 file
 # move required binaries and libraries to clean container
 COPY --from=build /usr/lib/libndctl* /usr/lib/
```

I ended up making some other changes:
- spelling fix ("Pmem" -> "PMEM")
- &&
- also clean up the swupd layer in the build image, because the content gets cached (smaller is better)
- single line RUN commands (IMHO more readable this way and the lines weren't longer than the others)

okartau commented:
yes, looks good

avalluri commented on cmd/pmem-csi-driver/Dockerfile:
```diff
@@ -17,19 +18,31 @@ RUN make install
 LABEL description="Pmem CSI Driver"
-RUN apk add --update kmod eudev util-linux libuuid e2fsprogs xfsprogs lvm2 file
+RUN swupd update; \
+    swupd bundle-add binutils file xfsprogs storage-utils; \
```
> Some bundles are obvious (xfsprogs and storage-utils), for one I can guess the purpose although I am not sure (file), and one is a mystery (binutils).

`binutils` is not needed, can be removed. I will update PR.
> 
> For each of them (even the obvious ones), can you list why they are needed? That will make it possible to re-evaluate this in the future.

Where should I list, in commit message?

avalluri replied:
> Use `&&` instead of `;`.

Done

avalluri commented:
@pohly I updated PR with your (docker --pull, &&, Pmem -> PMEM) changes. I did not take `CACHEBUST` change, as I think it's not required. And even I removed `binutils` from `swupd bundle-add` list in the `pmem-csi-driver` image. Now the sizes seem consistent.

```
192.168.8.1:5000/pmem-vgm                              canary              6e183e89011b        2 minutes ago       508MB
192.168.8.1:5000/pmem-ns-init                          canary              156efa591e7a        5 minutes ago       192MB
192.168.8.1:5000/pmem-csi-driver                       canary              9b4b58de0685        7 minutes ago       525MB
```
The change in size of `pmem-csi-driver` is result of not installing `binutils`.

Can you have a look.

avalluri commented on cmd/pmem-csi-driver/Dockerfile:
```diff
@@ -1,14 +1,15 @@
 
 WORKDIR /
-RUN wget https://github.com/pmem/ndctl/archive/v${NDCTL_VERSION}.tar.gz
+RUN curl -LO https://github.com/pmem/ndctl/archive/v${NDCTL_VERSION}.tar.gz
```
Done

avalluri commented on cmd/pmem-ns-init/Dockerfile:
```diff
@@ -1,14 +1,15 @@
 
 WORKDIR /
-RUN wget https://github.com/pmem/ndctl/archive/v${NDCTL_VERSION}.tar.gz
+RUN curl -LO https://github.com/pmem/ndctl/archive/v${NDCTL_VERSION}.tar.gz
```
Done

avalluri commented on cmd/pmem-ns-init/Dockerfile:
```diff
@@ -18,18 +19,24 @@ RUN make install
 # build clean container
-FROM golang:alpine
+FROM clearlinux:base
+RUN swupd update; \
```
Done

avalluri commented on cmd/pmem-vgm/Dockerfile:
```diff
@@ -1,14 +1,15 @@
 
 WORKDIR /
-RUN wget https://github.com/pmem/ndctl/archive/v${NDCTL_VERSION}.tar.gz
+RUN curl -LO https://github.com/pmem/ndctl/archive/v${NDCTL_VERSION}.tar.gz
```
Done

pohly commented:
Amarnath Valluri <notifications@github.com> writes:
>> For each of them (even the obvious ones), can you list why they are needed? That will make it possible to re-evaluate this in the future.
>
> Where should I list, in commit message?

In the code. The more information is there, the easier it is to find
when people start looking for it.

avalluri commented:
> Amarnath Valluri <notifications@github.com> writes:
> > For each of them (even the obvious ones), can you list why they are needed? That will make it possible to re-evaluate this in the future. Where should I list, in commit message?
> In the code. The more information is there, the easier it is to find when people start looking for it.

Done.
